zenith_utils/src/pq_proto.rs

```diff
@@ -169,44 +172,60 @@ impl FeStartupMessage {
         }
 
-        let version = stream.read_u32::<BE>()?;
-        let kind = match version {
-            CANCEL_REQUEST_CODE => StartupRequestCode::Cancel,
-            NEGOTIATE_SSL_CODE => StartupRequestCode::NegotiateSsl,
-            NEGOTIATE_GSS_CODE => StartupRequestCode::NegotiateGss,
-            _ => StartupRequestCode::Normal,
-        };
+        let request_code = stream.read_u32::<BE>()?;
 
         // the rest of startup packet are params
         let params_len = len - 8;
         let mut params_bytes = vec![0u8; params_len];
         stream.read_exact(params_bytes.as_mut())?;
 
-        // Then null-terminated (String) pairs of param name / param value go.
-        let params_str = str::from_utf8(&params_bytes).unwrap();
-        let params = params_str.split('\0');
-        let mut params_hash: HashMap<String, String> = HashMap::new();
-        for pair in params.collect::<Vec<_>>().chunks_exact(2) {
-            let name = pair[0];
-            let value = pair[1];
-            if name == "options" {
-                // deprecated way of passing params as cmd line args
-                for cmdopt in value.split(' ') {
-                    let nameval: Vec<&str> = cmdopt.split('=').collect();
-                    if nameval.len() == 2 {
-                        params_hash.insert(nameval[0].to_string(), nameval[1].to_string());
+        // Parse params depending on request code
+        let most_sig_16_bits = request_code >> 16;
+        let least_sig_16_bits = request_code & ((1 << 16) - 1);
+        let message = match (most_sig_16_bits, least_sig_16_bits) {
+            (RESERVED_INVALID_MAJOR_VERSION, CANCEL_REQUEST_CODE) => {
+                ensure!(params_len == 8, "expected 8 bytes for CancelRequest params");
+                let mut cursor = Cursor::new(params_bytes);
+                FeInitialMessage::CancelRequest(CancelKeyData {
+                    backend_pid: cursor.read_i32::<BigEndian>()?,
+                    cancel_key: cursor.read_i32::<BigEndian>()?,
+                })
+            }
+            (RESERVED_INVALID_MAJOR_VERSION, NEGOTIATE_SSL_CODE) => FeInitialMessage::SSLRequest,
+            (RESERVED_INVALID_MAJOR_VERSION, NEGOTIATE_GSS_CODE) => FeInitialMessage::GSSENCRequest,
+            (RESERVED_INVALID_MAJOR_VERSION, unrecognized_code) => {
+                bail!("Unrecognized request code {}", unrecognized_code)
+            }
+            (major_version, minor_version) => {
+                // TODO bail if protocol major_version is not 3?
+                // Parse null-terminated (String) pairs of param name / param value
+                let params_str = str::from_utf8(&params_bytes).unwrap();
+                let mut params_tokens = params_str.split('\0');
+                let mut params: HashMap<String, String> = HashMap::new();
+                while let Some(name) = params_tokens.next() {
+                    let value = params_tokens.next().ok_or_else(|| {
```

Comment:
We already have itertools in Cargo.lock, which has a nice method [`tuple_windows`](https://docs.rs/itertools/0.10.3/itertools/trait.Itertools.html#method.tuple_windows). Perhaps you could use it.

Reply:
And by the way, there's also https://docs.rs/itertools/0.10.3/itertools/trait.Itertools.html#method.next_tuple, which means we'll keep an iterator in case we wanted to check for leftovers.